### PR TITLE
Change wallet `currentTip` from `SlotId` to `BlockHeader`

### PR DESCRIPTION
**Pull request by Anviking**
# Issue Number

#219 

# Overview

- [x] Change the tip from `SlotId` to `BlockHeader` 

# Comments

- Remake of #366 
- The gist of it should be here, but I'll go through and polish / self-review it
- I have a concern about if we want to extend `BlockHeader` in the future
- We _could_ still parameterise over the tip (but even if we solve the Sqlite schema by storing `Either tip1 tip2` as `(Maybe tip1, Maybe tip2)`, we introduce a new dimension of complexity in the database, which might be undesired)

### CODE REVIEW

**Comment by Anviking** [comment]
<s>This will never work 😕 

`BlockHeader` stores the blockId of the parent. So if the tip is `n` we must call `getDescendants` on (n-1). This was fine, I thought — but of course we can't ask for the descendants of block `-1` to fetch block `0`.

But, I think we can't go wrong with `data Cursor = { blockId :: Hash "BlockHeader", header :: BlockHeader }`, which also happens to be the exact type used for the current `networkTip`.</s>

File: lib/core/src/Cardano/Wallet/Primitive/Model.hs [hunk 1]
```diff
@@ -157,9 +157,15 @@ instance NFData (Wallet s t) where
...
+    => BlockHeader
+    -- ^ Very first 'BlockHeader'
+    -> s
     -> Wallet s t
-initWallet = Wallet mempty mempty (SlotId 0 0)
+initWallet = Wallet mempty mempty genesisBlockHeader
+  where
+    genesisBlockHeader = BlockHeader
+        (SlotId 0 0)
+        (Hash "\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL")
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Primitive/Model.hs]: I think this was wrongly merged / cherry-picked ^^^

**Review by KtorZ — APPROVED** [review]
:+1:

**Comment by KtorZ** [comment]
rebased on top of master + improved comment for `defaultRetryPolicy` as suggested in #400

**Comment by Anviking** [comment]
Squashed relevant commits into `Change type of tip from SlotId to BlockHeader`.

**Comment by KtorZ** [comment]
@Anviking Not so sure about the squash here. I think some commits deserved to be separated for readability. Styling stuff and indentation fixes, sure; but structural changes were better off separated IMO.

**Pull request merged**